UI: Add cluster status warning before provisioning the volume in Prometheus

ChengYanJin opened the pull request:
**Component**: ui/monitoring

**Context**:  
As a TS user, I would like to know how to deal with the error status of the cluster(no volume provisioned for Prometheus pod) instead of the red state of Prometheus.

**Summary**:

- Add the `Unknown` status for the cluster when there isn't provision volume for Prometheus. 

- Add a warning banner to explain the status `Prometheus is not available.` and how to deal with the issue which should refer the position of the doc.

**Acceptance criteria**: 
In order to reproduce the issue, you can simply comment on the following line in `Vagrantfile` and then do `vagrant up` so that we won't provision any volume during deploy the bootstrap node.
```
machine.vm.provision "create-volumes",
type: "shell",
inline: CREATE_VOLUMES
```
When you go to the monitoring page, you will wait a moment(until the api timeout) and get the following page.
![image](https://user-images.githubusercontent.com/18453133/68218040-fbe10d00-ffe3-11e9-8c2d-aed84e02e68e.png)

Closes: #1890

ChengYanJin committed:
ui/monitoring: Add unknown status for cluster before provisioning the volume in Prometheus pod

Instead of displaying the red status of cluster, we diplay unknown in cluster status and add a banner to guide users.

Refs: #1890

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

ChengYanJin committed:
ui/monitoring: update the test of monitoring

Update the test and add one case when there is no provision volume for Prometheus pod.

Refs: #1890

carlito-scality commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -121,14 +128,32 @@ export function* fetchClusterStatus() {
+    if (!prometheusPod.error) {
+      const prometheusPodCondition =
+        prometheusPod?.body?.items[0]?.status?.conditions[0]?.type;
+      if (prometheusPodCondition === 'PodScheduled') {
```
Why "PodScheduled" instead of "Ready"? You should also check the status of the condition

ChengYanJin replied:
The status of the condition is `False` so I think it's better to check the type of condition?

gdemonet replied:
Mmh, actually this is not super robust. If one day another `condition` is added to _Pod_ `status`es, this will become somehow flaky. Instead, try something like:

```javascript
const conditions = prometheusPod?.body?.items[0]?.status?.conditions;
const scheduledCondition = conditions?.find(c => c.type === 'PodScheduled');
if (!scheduledCondition.status) {
  // update state with `scheduledCondition.message`
} else {
  // that's an unknown situation, handling TBD
}
```

carlito-scality replied:
Indeed, you have a status for each type of the condition. you should filter `prometheusPod?.body?.items[0]?.status?.conditions` to get the "Ready" (or "PodScheduled" ?)condition instead of getting the first condition. Then check the status of this condition if it is `false` then store the message

carlito-scality commented on ui/src/containers/ClusterMonitoring.js:
```diff
@@ -238,7 +255,22 @@ const ClusterMonitoring = props => {
       </ClusterStatusTitleContainer>
-
+      {clusterStatus.value === CLUSTER_STATUS_UNKNOWN &&
+      cluster.podStatusMessage ===
```
This check is not robust IMO. @gdemonet any ideas?

carlito-scality commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -20,17 +22,18 @@ const defaultState = {
     kubeSchedulerStatus: 0,
     kubeControllerManagerStatus: 0,
     error: null,
+    podStatusMessage: null,
```
you should rename it to "prometheusPodStatus" instead to be clearer. and It should contain the status of the pod instead of only the message

ChengYanJin replied:
I will change the name from `podStatusMessage` to `prometheusPodStatusMessage`
Not sure if it's useful to store the status of the pod as in this case it's `pending`

gdemonet replied:
IMO, this kind of information would make more sense in the `monitoring` Redux state tree, not in `monitoring.cluster`. But we can move this to a debt ticket, as it's not critical.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality

ChengYanJin committed:
ui/monitoring: Add translation for banner

Refs: #1890

gdemonet commented on ui/src/containers/ClusterMonitoring.js:
```diff
@@ -238,7 +255,16 @@ const ClusterMonitoring = props => {
         </RightClusterStatusContainer>
       </ClusterStatusTitleContainer>
-
+      {clusterStatus.value === CLUSTER_STATUS_UNKNOWN &&
```
Instead of checking this `UNKNOWN` status, why not use the `monitoring.isPrometheusApiUp` boolean that already exists in the state?

`UNKNOWN` could exist while a query is in progress, for instance.

gdemonet commented on ui/src/containers/ClusterMonitoring.js:
```diff
@@ -238,7 +255,16 @@ const ClusterMonitoring = props => {
       </ClusterStatusTitleContainer>
-
+      {clusterStatus.value === CLUSTER_STATUS_UNKNOWN &&
+      cluster.prometheusPodStatusMessage ===
```
Indeed, I'm not sure if this value will always match this string. Let's be slightly more flexible:

```javascript
cluster.prometheusPodStatusMessage.includes(`didn't find available persistent volumes to bind`)
```

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -20,17 +22,18 @@ const defaultState = {
+    prometheusPodStatusMessage: null,
     isLoading: false,
-    isRefreshing: false
+    isRefreshing: false,
```
What's the difference between `loading` and `refreshing`? Confuses me, though not related to this PR.

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -20,17 +22,18 @@ const defaultState = {
     kubeSchedulerStatus: 0,
     kubeControllerManagerStatus: 0,
     error: null,
+    prometheusPodStatusMessage: null,
```
IMHO, this should not be in `monitoring.cluster`, simply `monitoring`, or maybe even go as far as:

```
const defaultState = {
  // ...
  cluster: { ... },
  prometheus: {
    isAvailable: false, // the old `isPrometheusApiUp`
    podStatusMessage: null,
  },
}
```

That could help keep things organized. But feel free to put this in a debt ticket.

ChengYanJin replied:
> IMHO, this should not be in `monitoring.cluster`, simply `monitoring`, or maybe even go as far as:
> 
> ```
> const defaultState = {
>   // ...
>   cluster: { ... },
>   prometheus: {
>     isAvailable: false, // the old `isPrometheusApiUp`
>     podStatusMessage: null,
>   },
> }
> ```
> 
> That could help keep things organized. But feel free to put this in a debt ticket.

Will use`isPrometheusVolumeProvisioned` prop in cluster to control the banner visibility.

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -121,14 +128,31 @@ export function* fetchClusterStatus() {
+      }
+    } else {
+      console.error('Cannot find the Prometheus pod.');
+    }
```
I would move this logic in `handleClusterError` (which could be renamed `handlePrometheusError`, for clarity).

ChengYanJin replied:
> I would move this logic in `handleClusterError` (which could be renamed `handlePrometheusError`, for clarity).

Refactoring the handlePrometheusError method in a separate commit

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -121,14 +128,31 @@ export function* fetchClusterStatus() {
+        clusterHealth.prometheusPodStatusMessage =
+          prometheusPod?.body?.items[0]?.status?.conditions[0]?.message;
+      } else {
+        console.error('This is an error in the cluster.');
```
Not sure it's good practice to print error messages in the console. Also, this message doesn't help understand the actual state or why we hit it.

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -121,14 +128,31 @@ export function* fetchClusterStatus() {
+        console.error('This is an error in the cluster.');
+      }
+    } else {
+      console.error('Cannot find the Prometheus pod.');
```
This could be due to another error, you should check the HTTP status code is `404` before saying that. Anyway (as said above), I don't think we should print error messages in the console.

ChengYanJin replied:
> This could be due to another error, you should check the HTTP status code is `404` before saying that. Anyway (as said above), I don't think we should print error messages in the console.

Removed the console.error!
Set clusterHealth.error='Unkown' instead of.

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -121,14 +128,31 @@ export function* fetchClusterStatus() {
+    const prometheusPod = yield call(
+      ApiK8s.getPodInNamespace,
+      'metalk8s-monitoring',
+      'prometheus',
```
Is this supposed to be the _Pod_ `metadata.name`? If it is, this won't work, as the actual _Pod_ is created through a _StatefulSet_. If it isn't, then the `getPodInNamespace` method is confusing (but that's debt, not your PR).

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -90,6 +93,9 @@ export function* handleClusterError(clusterHealth, result) {
     clusterHealth.error = `Prometheus - ${result.error.response.statusText}`;
+  } else if (clusterHealth.prometheusPodStatusMessage) {
+    yield put(setPrometheusApiAvailable(false));
+    clusterHealth.error = 'Unknown';
```
No, actually if you want to have a different state in `clusterHealth` in case there's a missing Volume for Prometheus, then it's not "unknown", it's "missing a Volume". But I'd argue this can remain `prometheus_unavailable`, and the banner is enough to explain to the user **why** it is unavailable.

carlito-scality replied:
it should be 'unknown' instead of "Unknown" to be translated

gdemonet commented on ui/src/services/k8s/api.js:
```diff
@@ -183,3 +183,18 @@ export async function createEnvironment(body) {
+  } catch (error) {
+    return { error };
+  }
+}
```
Oh, that's not a good name/signature for this method IMHO (for instance, it's a `list` action, not a `get`, since you retrieve multiple objects that match a given `labelSelector`). But let's move it to a debt ticket, I think there are other methods like this one that could be improved.

ChengYanJin replied:
> Oh, that's not a good name/signature for this method IMHO (for instance, it's a `list` action, not a `get`, since you retrieve multiple objects that match a given `labelSelector`). But let's move it to a debt ticket, I think there are other methods like this one that could be improved.

Make sense!
Will rename the method to `queryPodInNameSpace` maybe?

gdemonet reviewed (changes requested):
Numerous comments, but let's focus on the actual handling of the message to merge this PR (keep the same `clusterHealth` that says "Prometheus is not available", and rely on the Pod status to display the banenr).

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality

* @gdemonet

carlito-scality commented on ui/src/containers/ClusterMonitoring.js:
```diff
@@ -275,12 +300,14 @@ const makeClusterStatus = (state, props) => {
     label = intl.messages[cluster.error] || cluster.error;
   }
-
+  if (cluster.prometheusPodStatusMessage) {
```
Here you use `prometheusPodStatusMessage ` message to compute the status of the cluster. It makes more sense to use the status of the prometheusPod instead. So after calling ApiK8s.getPodInNamespace, store the status in clusterHealth and use it here and when setting ` clusterHealth.error = 'Unknown';` in the saga

ChengYanJin replied:
I use `isPrometheusVolumeProvisioned` to control the banner visibility and store it in clusterHealth. Is that make sense to you？

ChengYanJin committed:
ui/monitoring: Refactoring handleClusterError method

Rename handleClusterError to handlePrometheusError as we already known it's Prometheus issue

Refs: #1890

ChengYanJin committed:
ui/monitoring: Add isPrometheusVolumeProvisioned in cluster saga

when isPrometheusVolumeProvisioned is false we display the banner

Refs: #1890

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -86,13 +89,37 @@ function getClusterQueryStatus(result) {
+    } else {
+      yield put(setPrometheusApiAvailable(false));
+      clusterHealth.error = 'prometheus_unavailable';
+    }
```
I don't get why we still have an `unknown` status. `prometheus_unavailable` should be enough:

```javascript
// Check if the error is due to an unschedulable Pod, because of missing PV
const prometheusPod = yield call(
  ApiK8s.queryPodInNamespace,
  'metalk8s-monitoring',
  'prometheus',
);

if (!prometheusPod.error) {
  const conditions = prometheusPod?.body?.items[0]?.status?.conditions;
  const scheduledCondition = conditions?.find(
    c => c.type === 'PodScheduled',
  );
  if (
    scheduledCondition.message.includes(
      `didn't find available persistent volumes to bind`,
    )
  ) {
    clusterHealth.isPrometheusVolumeProvisioned = false;
  }
}

// In any case, set these values
yield put(setPrometheusApiAvailable(false));
clusterHealth.error = 'prometheus_unavailable';
```

gdemonet commented on ui/src/translations/en.json:
```diff
@@ -124,5 +124,7 @@
-  "advanced_monitoring": "Advanced Monitoring"
+  "advanced_monitoring": "Advanced Monitoring",
+  "prometheus_not_available":"Prometheus is not available.",
+  "provision_prometheus_volume":"Please refer to MetalK8s Quickstart Guide > Deployment of the Bootstrap node > Installation > Provision storage for Prometheus services"
```
We could just translate `Please refer to`, and leave the rest untranslated (since the doc is in English anyway).

gdemonet reviewed (changes requested):
Last changes, I promise ;)

gdemonet commented:
By the way, this PR will probably conflict with #1979 (since it changes some translations as well), so it may make sense to rebase as soon as it gets merged.

gdemonet commented on ui/src/containers/ClusterMonitoring.js:
```diff
@@ -275,12 +305,14 @@ const makeClusterStatus = (state, props) => {
     label = intl.messages[cluster.error] || cluster.error;
   }
-
+  if (!cluster.isPrometheusVolumeProvisioned) {
```
Almost, that should be:

```suggestion
  if (!state.app.monitoring.isPrometheusApiUp) {
```

ChengYanJin commented:
> By the way, this PR will probably conflict with #1979 (since it changes some translations as well), so it may make sense to rebase as soon as it gets merged.

Thanks for the review!!!

gdemonet reviewed (dismissed):
LGTM! Thanks for the reactivity ;)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

ChengYanJin commented:
/approve

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/1890-give-warning-in-ui-when-storage-is-not-yet-provisioned` and the
destination branch `development/2.4`.

Please resolve the conflict on **the feature branch** (`improvement/1890-give-warning-in-ui-when-storage-is-not-yet-provisioned`).

```bash
 $ git fetch
 $ git checkout origin/improvement/1890-give-warning-in-ui-when-storage-is-not-yet-provisioned
 $ git merge origin/development/2.4
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/1890-give-warning-in-ui-when-storage-is-not-yet-provisioned
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

carlito-scality commented on ui/src/containers/ClusterMonitoring.js:
```diff
@@ -182,12 +200,14 @@ const ClusterMonitoring = props => {
   const sortedAlerts = sortSelector(alertsList, sortBy, sortDirection);
 
-  return (
+  return alerts.isLoading ? (
```
You should not use `alerts.isLoading` here otherwise the whole page will refresh every 20s. Let's remove it for now and will improve it later.

ChengYanJin replied:
ok. So I removed the `<Loader/>` for the moment Will find a better solution.

carlito-scality reviewed (changes requested):
One last request ;)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality







*The following options are set:* **approve**

ChengYanJin commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/575955097a538a1adafdff75fb094d50520800cb) did not succeed in branch improvement/1890-give-warning-in-ui-when-storage-is-not-yet-provisioned.




*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/5f9cf95065d96b7d9015622cdb841358aa7f5b34) did not succeed in branch improvement/1890-give-warning-in-ui-when-storage-is-not-yet-provisioned.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye chengyanjin.
